src/yafscii.1
.\" Automatically generated by Pod::Man 2.23 (Pod::Simple 3.14)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "YAFSCII 1"
.TH YAFSCII 1 " 3-Mar-2013" "0.9.0" "QoF"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
yafscii \- YAF Flow Printer
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 7
\&    yafscii [\-\-in INPUT_SPECIFIER] [\-\-out OUTPUT_SPECIFIER]
\&            [\-\-nextdir PROCESSED_INPUT_DIRECTORY]
\&            [\-\-faildir FAILED_INPUT_DIRECTORY]
\&            [\-\-poll POLLING_DELAY] [\-\-lock]
\&            [\-\-log LOG_SPECIFIER] [\-\-loglevel LOG_LEVEL]
\&            [\-\-verbose] [\-\-version] [\-\-daemon] [\-\-foreground]
\&            [\-\-tabular] [\-\-mac] [\-\-print\-header]
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fByafscii\fR takes \s-1IPFIX\s0 flow data files generated by \fIyaf\fR\|(1) and prints them 
in an \s-1ASCII\s0 format loosely analogous to that produced by \fItcpdump\fR\|(1), with one 
flow per line. The text output format is detailed in the \fB\s-1OUTPUT\s0\fR section, 
below. \fByafscii\fR is generally intended to be used to print single files for 
verification or debugging purposes, or to operate in a pipe with \fIyaf\fR\|(1),
but it can be used as a daemon as well.  \fByafscii\fR ignores \fIyaf\fR\|(1) stats records.
.SH "OPTIONS"
.IX Header "OPTIONS"
.SS "Input Options"
.IX Subsection "Input Options"
The input specifier determines where yafscii will read its 
input from. If the input specifier is not given, yaf defaults to reading 
from standard input.
.IP "\fB\-\-in\fR \fI\s-1INPUT_SPECIFIER\s0\fR" 4
.IX Item "--in INPUT_SPECIFIER"
\&\fI\s-1INPUT_SPECIFIER\s0\fR is an input specifier. This is a filename, a directory 
name, a file glob pattern (in which case it should be 
escaped or quoted to prevent the shell from expanding the glob pattern), or 
the string \fB\-\fR to read from standard input.
.SS "Output Options"
.IX Subsection "Output Options"
The output specifier determines where yaf will send its output. If reading 
standard input, output defaults to standard output. If reading from files on 
disk, output defaults to one file per input file, named as the input file in 
the same directory as the input file with a \fB.txt\fR extension.
.IP "\fB\-\-out\fR \fI\s-1OUTPUT_SPECIFIER\s0\fR" 8
.IX Item "--out OUTPUT_SPECIFIER"
\&\fI\s-1OUTPUT_SPECIFIER\s0\fR is an output specifier. This should be a filename or a 
directory name, or the string \fB\-\fR to write to standard output.
.IP "\fB\-\-tabular\fR" 8
.IX Item "--tabular"
Use tabular output mode, which is designed for easy parsability over
human readability. See the Tabular Output section below for details.
.IP "\fB\-\-mac\fR" 8
.IX Item "--mac"
Used with \fB\-\-tabular\fR mode to print source and destination \s-1MAC\s0 Addresses.
.IP "\fB\-\-print\-header\fR" 8
.IX Item "--print-header"
Used with \fB\-\-tabular\fR mode to print column headers.
.SS "Daemon Options"
.IX Subsection "Daemon Options"
These options are used to run yafscii in daemon mode for batch processing of
pcap dumpfiles.
.IP "\fB\-\-daemon\fR" 4
.IX Item "--daemon"
Run yafscii in daemon mode. Instead of processing its input then exiting, 
yafscii will continually look for new input matching its input specifier. 
This will cause yaf to fork into the background and exit.
=item \fB\-\-foreground\fR
.Sp
Instead of forking in \fB\-\-daemon\fR mode, stay in the foreground. 
Useful for debugging.
.IP "\fB\-\-lock\fR" 4
.IX Item "--lock"
Use lockfiles for concurrent file access protection. Highly recommended in 
\&\fB\-\-daemon\fR mode, especially if two daemons are interacting through a 
given directory.
.IP "\fB\-\-poll\fR \fI\s-1POLLING_DELAY\s0\fR" 4
.IX Item "--poll POLLING_DELAY"
\&\fI\s-1POLLING_DELAY\s0\fR is the polling delay in seconds; how long yaf will 
wait for new input when none is available. The default is 60 seconds.
.IP "\fB\-\-nextdir\fR \fI\s-1PROCESSED_INPUT_DIRECTORY\s0\fR" 4
.IX Item "--nextdir PROCESSED_INPUT_DIRECTORY"
When reading from files, if this option is present, input files will be 
moved to \fI\s-1PROCESSED_INPUT_DIRECTORY\s0\fR after they are successfully processed. 
The special string \fBdelete\fR will cause successfully processed input to be 
removed instead. This option is required in daemon mode.
.IP "\fB\-\-faildir\fR \fI\s-1FAILED_INPUT_DIRECTORY\s0\fR" 4
.IX Item "--faildir FAILED_INPUT_DIRECTORY"
When reading from files, if this option is present, input files will be 
moved to \fI\s-1FAILED_INPUT_DIRECTORY\s0\fR if processing failed. 
The special string \fBdelete\fR will cause failed input to be 
removed instead. This option is required in daemon mode.
.SS "Logging Options"
.IX Subsection "Logging Options"
These options are used to specify how log messages are routed. yaf can 
log to standard error, regular files, or the \s-1UNIX\s0 syslog facility.
.IP "\fB\-\-log\fR \fI\s-1LOG_SPECIFIER\s0\fR" 4
.IX Item "--log LOG_SPECIFIER"
Specifies destination for log messages. \fI\s-1LOG_SPECIFIER\s0\fR can be a \fIsyslog\fR\|(3)
facility name, the special value \fBstderr\fR for standard error, or the 
\&\fIabsolute\fR path to a file for file logging. Standard error logging is only 
available in \fB\-\-daemon\fR mode if \fB\-\-foreground\fR is present. The default log 
specifier is \fBstderr\fR if available, \fBuser\fR otherwise.
.IP "\fB\-\-loglevel\fR \fI\s-1LOG_LEVEL\s0\fR" 4
.IX Item "--loglevel LOG_LEVEL"
Specify minimum level for logged messages. In increasing levels of verbosity, 
the supported log levels are \fBquiet\fR, \fBerror\fR, \fBcritical\fR, \fBwarning\fR, 
\&\fBmessage\fR, \fBinfo\fR, and \fBdebug\fR. The default logging level is \fBwarning\fR.
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
Equivalent to \fB\-\-loglevel debug\fR.
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
If present, print version and copyright information to standard error and exit.
.SH "OUTPUT"
.IX Header "OUTPUT"
.SS "Human-Readable Output"
.IX Subsection "Human-Readable Output"
yafscii's default output format, like that of tcpdump, is designed to be easily 
human-readable, at the possible expense of ease of automated parsing. 
Each flow is represented by a single output line representing the flow 
itself, followed by zero or more indented lines containing flow payload in 
hexdump format. This section details each flow format, where each field 
specifier is as follows:
.IP "\fIstart-time\fR, \fIend-time\fR" 4
.IX Item "start-time, end-time"
Flow start or end time in \s-1ISO\s0 8601 format, with milliseconds 
(\fI\s-1YYYY\s0\fR\-\fI\s-1MM\s0\fR\-\fI\s-1DD\s0\fR \fIhh\fR:\fImm\fR:\fIss\fR.\fIssss\fR). Start time is printed with
a date; end time is not. End time is only present if the flow has a
non-zero duration.
.IP "\fIduration\fR" 4
.IX Item "duration"
Flow duration in fractional seconds. Only present if the flow has a 
non-zero duration.
.IP "\fIproto\fR" 4
.IX Item "proto"
\&\s-1IP\s0 protocol identifier in decimal format.
.IP "\fIsip\fR, \fIdip\fR" 4
.IX Item "sip, dip"
Source or destination IPv4 address in dotted-quad format or IPv6 address in 
\&\s-1RFC\s0 2373 format.
.IP "\fIsp\fR, \fIdp\fR" 4
.IX Item "sp, dp"
Source or destination transport port in decimal format.
.IP "\fItype\fR, \fIcode\fR" 4
.IX Item "type, code"
\&\s-1ICMP\s0 type or code in decimal format.
.IP "\fIisn\fR, \fIrisn\fR" 4
.IX Item "isn, risn"
Forward or reverse initial \s-1TCP\s0 sequence number in hexadecimal format.
.IP "\fIiflags\fR, \fIriflags\fR, \fIuflags\fR, \fIruflags\fR" 4
.IX Item "iflags, riflags, uflags, ruflags"
Foward or reverse first-packet \s-1TCP\s0 flags; forward or reverse \fIn\fRth-packet 
\&\s-1TCP\s0 flags union; where each flags bit is represented by the first character 
in the flag's name: \fBF\fR\s-1IN\s0, \fBS\fR\s-1YN\s0, \fBR\fR\s-1ST\s0, \fBP\fR\s-1SH\s0, \fBA\fR\s-1CK\s0, \fBU\fR\s-1RG\s0, \fBE\fR\s-1CE\s0, 
\&\fBC\fR\s-1WR\s0. The character \fB0\fR means no flags are set (and will appear in the 
\&\fIn\fRth-packet field for single-packet \s-1TCP\s0 flows).
.IP "\fItag\fR, \fIrtag\fR" 4
.IX Item "tag, rtag"
Forward or reverse first-packet 802.1q \s-1VLAN\s0 tag in hexadecimal format.
.IP "\fIsrcMacAddress\fR, \fIdestMacAddress\fR" 4
.IX Item "srcMacAddress, destMacAddress"
Source or Destination \s-1MAC\s0 Address.
.IP "\fIpkt\fR, \fIrpkt\fR" 4
.IX Item "pkt, rpkt"
Forward or reverse packet count in decimal format.
.IP "\fIoct\fR, \fIroct\fR" 4
.IX Item "oct, roct"
Forward or reverse octet count in decimal format.
.IP "\fIrtt\fR" 4
.IX Item "rtt"
Round-trip time estimate in milliseconds in decimal format.
.IP "\fIend-reason\fR" 4
.IX Item "end-reason"
If not present, the flow ended normally (i.e., by \s-1TCP\s0 \s-1RST\s0 or \s-1FIN\s0). Otherwise, the \fIend-reason\fR is one of the following strings:
.RS 4
.IP "\fBidle\fR" 5
.IX Item "idle"
Flow was expired by idle timeout. No packets were received for 
\&\fI\s-1IDLE_TIMEOUT\s0\fR seconds (see \fIyaf\fR\|(1)) and the flow was presumed closed.
.IP "\fBactive\fR" 5
.IX Item "active"
Flow was expired by active timeout. The flow's duration was longer than 
\&\fI\s-1ACTIVE_TIMEOUT\s0\fR seconds (see \fIyaf\fR\|(1)) and the flow was flushed from the 
flow table.
.IP "\fBeof\fR" 5
.IX Item "eof"
Flow was still active in the flow table at the end of the dumpfile or at 
\&\fIyaf\fR\|(1) shutdown time; it was flushed as the flow table was cleared.
.IP "\fBrsrc\fR" 5
.IX Item "rsrc"
Flow was prematurely flushed as \fBidle\fR because more than \fI\s-1FLOW_TABLE_MAX\s0\fR 
flows (see \fIyaf\fR\|(1)) were active in the flow table.
.IP "\fBforce\fR" 5
.IX Item "force"
\&\fByaf\fR forced a write of the flow, but the flow remained open.  This is only seen if yaf operated with the \fB\-\-udp\-uniflow\fR flag, which exports each \s-1UDP\s0 packet as a flow record, but allows the flow to remain open until it closes naturally by idle and active timeouts (see \fIyaf\fR\|(1)).
.RE
.RS 4
.RE
.IP "\fBapplabel\fR" 4
.IX Item "applabel"
The application label, if \fIyaf\fR\|(1) was built with application labeling enabled and
the application labeler was able to identify the payload in the flow.
.IP "\fBentropy\fR" 4
.IX Item "entropy"
The Shannon-Fano Entropy for the forward then the reverse flow payload if the payload
existed and \fIyaf\fR\|(1) was built with entropy enabled.
.PP
Each flow line format is as follows:
.IP "Unidirectional \s-1IP\s0 flow" 4
.IX Item "Unidirectional IP flow"

NEWS
Version 0.9: 2013-xx-xx
=======================



Release notes for YAF releases prior to the forking of QoF follow.

Version 2.3.2: 2012-09-14
=========================

Bug Fix to maintain compatibility with older versions of GLib and libpcap

Version 2.3.1: 2012-09-10
=========================

DPI Improvements

Additional Pcap Export Option --index-pcap

Add option to manually set ingress/egress interface fields

Add tool to create pcap from pcap metafile

Bug Fixes

Version 2.3.0: 2012-07-31
=========================

Added DHCP Fingerprinting Capability

Added ability to export DNSSEC information

Significant X.509 Certificate Capture and Export Enhancements

Added Bivio Interface Labeling

DPI Improvements

Added Enhanced Flow Attributes and Statistics Export

Added ability to index PCAP file

Added New Application Labels: MGCP, MEGACO

Bug Fixes

Version 2.2.2: 2012-03-30
=========================

Bug Fix for Vlan Tagging

Version 2.2.1: 2012-03-08
==========================

Bug Fixes

Version 2.2.0: 2012-02-29
============================

New Application Labels (MSNP, RTP, RTCP, Jabber)

Rolling Pcap output and pcap-per-flow options.

<a href="https://tools.netsa.cert.org/confluence/display/tt/p0f+fingerprints">CERT p0f Fingerprints</a> included.

New option to process out-of-sequence flows.

Several other bug fixes.

Version 2.1.2: 2011-09-23
=============================

Added new --plugin-conf switch for adding a configuration file to a plugin

Added new --p0f-fingerprints switch to give location of p0f fingerprint files

Bug Fixes

Version 2.1.1: 2011-08-11
=============================

Important bug fix for application labeling SSL plugin

Version 2.1.0: 2011-07-27
=============================

New Information Element exported in every flow record, flowAttributes (CERT PEN 6871, IE 40).

YAF now checks if a flow has fixed-size packets and exports this flag using the new flowAttributes Information Element (see <a href="yaf.html">yaf</a>)

Reset Application Label on UDP-uniflows for Deep Packet Inspection

Fixed yafscii invalid parameter bug that may have existed on certain platforms

Added VNC (RFB Protocol) application label

DPI Enhancements

FlowEndReason IPFIX field is now set to 31 for udp-uniflows

For Cygwin: Added support for getting the yaf config directory via the Windows Registry

Several other bug fixes

Version 2.0.2: 2011-06-13
==============================

Improvements with Reassembly of TCP Fragments.

Bug Fix for DNS Deep Packet Inspection.

--no-frag switch now works.

Bug Fix for expiring flows that exceed the idle timeout when reading from a file.

Added the ability to configure YAF with WinPCAP.

Version 2.0.1: 2011-05-23
==============================

Bug Fix for compile error with --enable-daginterface

Enhancement for SNMPv3 application labeler

Version 2.0.0: 2011-04-28
==============================

This version requires <a href="../fixbuf/index.html">libfixbuf-1.0.0</a> or greater.

Added Napatech Adapter Integration (requires libpcapexpress).

YAF now exports TCP, payload, finger printing, p0f, MAC, entropy, and DPI flow information within an IPFIX subTemplateMultiList data type. 

Added the ability to export YAF capture statistics using IPFIX Options Templates.

The --stats or --no-stats were added to configure YAF stats output.

Added the ability to define Spread group types to use Spread as a manifold for flow export based on application, port, protocol, version, or vlan.

Added New Application Labels: DHCP, AIM, SOCKS, SMB, SNMP, NETBIOS.

Added a time-out buffer flush function.

Added SSL Certificate Capture.

Added DNS Resource Record Parsing.

Added Deep Packet Inspection for the MySQL protocol.

The --silk switch will maintain compatibility with SiLK by not nesting TCP information in the subTemplateMultiList data type. 

Deep Packet Inspection elements are read from one configuration file.

Added the ability to create new DPI elements from configuration file.

Added UDP Export and Template Retransmission.

Many Bug fixes and other enhancements.

Version 1.3.2: 2011-02-03
=============================

Bug fix for dnsplugin.c

Minor bug fix for fingerprint exporting.

Version 1.3.1: 2010-10-06
==============================

Important bug fix for p0f or fpexport enabled code.

Fixed bug in DNS Application Labeling Decoder.

Removed machine learning code for future work.

Version 1.3.0: 2010-09-20
===============================

Vlan tags are now a part of the flow key.

Vlan tags are now always exported.

--mac flag exports MAC addresses.

Fixed bug in DNS Application Labeling Decoder.

Fixed bug in libp0f Makefile.

Added --print-header switch to yafscii for use with tabular mode to print column headers.

Added --mac switch to yafscii to support printing of MAC addresses in tabular mode.

Version 1.2.0: 2010-07-27
===============================

Spread support has been added into libfixbuf and YAF to allow publish subscribe distribution of YAF sensor output.

Plugin support has returned to YAF to support basic deep packet inspection (DPI) and application labeling (see <a href="yafdpi.html"> yafdpi </a>).

Added 9 new protocols to the application labeling feature (see <a href="applabel.html">applabel</a>).

Added ability for signature detection through the application labeling mechanism.

Added --udp-uniflow switch to capture each UDP packet on a set port and export the payload (for DNS dissector creation).

Added --udp-payload to concatenate and export payload up to the max-payload value.

DNS DPI can be restricted to Authoritative and NXDomain responses only via compile switches.

Enhanced payload capture for TCP streams with out-of-order SYN packets.

Fixed a bug in processing small (less than 64-packets) PCAP files.

Fixed IPv6 header options bug.

Fixed bug in parsing capability for strings longer than 80 columns.


src/yafscii.c
/*
 ** yafscii.c
 ** YAF flow printer
 **
 ** ------------------------------------------------------------------------
 ** Copyright (C) 2006-2012 Carnegie Mellon University. All Rights Reserved.
 ** ------------------------------------------------------------------------
 ** Authors: Brian Trammell
 ** ------------------------------------------------------------------------
 ** @OPENSOURCE_HEADER_START@
 ** Use of the YAF system and related source code is subject to the terms
 ** of the following licenses:
 **
 ** GNU Public License (GPL) Rights pursuant to Version 2, June 1991
 ** Government Purpose License Rights (GPLR) pursuant to DFARS 252.227.7013
 **
 ** NO WARRANTY
 **
 ** ANY INFORMATION, MATERIALS, SERVICES, INTELLECTUAL PROPERTY OR OTHER
 ** PROPERTY OR RIGHTS GRANTED OR PROVIDED BY CARNEGIE MELLON UNIVERSITY
 ** PURSUANT TO THIS LICENSE (HEREINAFTER THE "DELIVERABLES") ARE ON AN
 ** "AS-IS" BASIS. CARNEGIE MELLON UNIVERSITY MAKES NO WARRANTIES OF ANY
 ** KIND, EITHER EXPRESS OR IMPLIED AS TO ANY MATTER INCLUDING, BUT NOT
 ** LIMITED TO, WARRANTY OF FITNESS FOR A PARTICULAR PURPOSE,
 ** MERCHANTABILITY, INFORMATIONAL CONTENT, NONINFRINGEMENT, OR ERROR-FREE
 ** OPERATION. CARNEGIE MELLON UNIVERSITY SHALL NOT BE LIABLE FOR INDIRECT,
 ** SPECIAL OR CONSEQUENTIAL DAMAGES, SUCH AS LOSS OF PROFITS OR INABILITY
 ** TO USE SAID INTELLECTUAL PROPERTY, UNDER THIS LICENSE, REGARDLESS OF
 ** WHETHER SUCH PARTY WAS AWARE OF THE POSSIBILITY OF SUCH DAMAGES.
 ** LICENSEE AGREES THAT IT WILL NOT MAKE ANY WARRANTY ON BEHALF OF
 ** CARNEGIE MELLON UNIVERSITY, EXPRESS OR IMPLIED, TO ANY PERSON
 ** CONCERNING THE APPLICATION OF OR THE RESULTS TO BE OBTAINED WITH THE
 ** DELIVERABLES UNDER THIS LICENSE.
 **
 ** Licensee hereby agrees to defend, indemnify, and hold harmless Carnegie
 ** Mellon University, its trustees, officers, employees, and agents from
 ** all claims or demands made against them (and any related losses,
 ** expenses, or attorney's fees) arising out of, or relating to Licensee's
 ** and/or its sub licensees' negligent use or willful misuse of or
 ** negligent conduct or willful misconduct regarding the Software,
 ** facilities, or other rights or assistance granted by Carnegie Mellon
 ** University under this License, including, but not limited to, any
 ** claims of product liability, personal injury, death, damage to
 ** property, or violation of any laws or regulations.
 **
 ** Carnegie Mellon University Software Engineering Institute authored
 ** documents are sponsored by the U.S. Department of Defense under
 ** Contract FA8721-05-C-0003. Carnegie Mellon University retains
 ** copyrights in all material produced under this contract. The U.S.
 ** Government retains a non-exclusive, royalty-free license to publish or
 ** reproduce these documents, or allow others to do so, for U.S.
 ** Government purposes only pursuant to the copyright license under the
 ** contract clause at 252.227.7013.
 **
 ** @OPENSOURCE_HEADER_END@
 ** ------------------------------------------------------------------------
 */

#define _YAF_SOURCE_
#include <yaf/autoinc.h>
#include <airframe/mio.h>
#include <airframe/mio_config.h>
#include <airframe/mio_source_file.h>
#include <airframe/mio_sink_file.h>
#include <airframe/logconfig.h>
#include <airframe/daeconfig.h>
#include <airframe/airutil.h>
#include <airframe/privconfig.h>
#include <yaf/yafcore.h>

typedef struct ytContext_st {
    fBuf_t          *fbuf;
    yfFlow_t        flow;
} ytContext_t;

static uint32_t     yaft_flows = 0;

static gboolean     yaft_tabular = FALSE;
static gboolean     yaft_mac = FALSE;
static gboolean     yaft_first = TRUE;
static gboolean     yaft_print_header = FALSE;

static uint32_t     yaft_cliflags = MIO_F_CLI_FILE_IN |
                                    MIO_F_CLI_DIR_IN |
                                    MIO_F_CLI_DEF_STDIN |
                                    MIO_F_CLI_FILE_OUT |
                                    MIO_F_CLI_DIR_OUT |
                                    MIO_F_CLI_DEF_STDOUT;

AirOptionEntry yaft_optentries[] = {
    AF_OPTION( "tabular", (char)0, 0, AF_OPT_TYPE_NONE, &yaft_tabular,
      "Print flows in tabular format", NULL ),
    AF_OPTION( "mac", (char)0, 0, AF_OPT_TYPE_NONE, &yaft_mac,
               "Print mac addresses (when used with --tabular)", NULL ),
    AF_OPTION( "print-header", (char)0, 0, AF_OPT_TYPE_NONE, &yaft_print_header,
               "Print column headers for tabular format", NULL),
    AF_OPTION_END
};

static void ytParseOptions(
    int             *argc,
    char            **argv[]) {

    AirOptionCtx *aoctx = NULL;

    aoctx = air_option_context_new("", argc, argv, yaft_optentries);

    mio_add_option_group(aoctx, yaft_cliflags);
    daec_add_option_group(aoctx);
    privc_add_option_group(aoctx);
    logc_add_option_group(aoctx, "yafscii", VERSION);

    air_option_context_set_help_enabled(aoctx);

    air_option_context_parse(aoctx);
    if (yaft_mac && !yaft_tabular) {
        g_warning("--mac requires --tabular");
    }
    if (yaft_print_header && !yaft_tabular) {
        g_warning("--print-header requires --tabular");
    }

    air_option_context_free(aoctx);
}

static gboolean ytOpenSource(
    MIOSource               *source,
    void                    *vctx,
    uint32_t                *flags,
    GError                  **err)
{
    ytContext_t          *yx = (ytContext_t *)vctx;

    /* start reading a YAF file */
    if (!(yx->fbuf = yfReaderForFP(yx->fbuf, mio_fp(source), err))) {
        *flags |= MIO_F_CTL_SOURCECLOSE;
        if (g_error_matches(*err, FB_ERROR_DOMAIN, FB_ERROR_EOF)) {
            g_clear_error(err);
        } else {
            *flags |= MIO_F_CTL_ERROR;
        }
        return FALSE;
    }

    return TRUE;
}

static gboolean ytProcess(
    MIOSource               *source,
    MIOSink                 *sink,
    void                    *vctx,
    uint32_t                *flags,
    GError                  **err)
{
    ytContext_t          *yx = (ytContext_t *)vctx;
    gboolean                ok;

    /* Print first line column headers if tabular */
    if (yaft_tabular && yaft_first && yaft_print_header) {
        yfPrintColumnHeaders(mio_fp(sink), yaft_mac, err);
        yaft_first = FALSE;
    }

    /* read a flow */
/*    if ((ok = yfReadFlowExtended(yx->fbuf, &(yx->flow), err))) { */
    if ((ok = yfReadFlow(yx->fbuf, &(yx->flow), err))) {
        ++yaft_flows;
        if (yaft_tabular) {
            ok = yfPrintDelimited(mio_fp(sink), &(yx->flow), yaft_mac, err);
        } else {
            ok = yfPrint(mio_fp(sink), &(yx->flow), err);
        }
    }

    /* handle error */
    if (!ok) {
        /* check for EOF */
        if (g_error_matches(*err, FB_ERROR_DOMAIN, FB_ERROR_EOF)) {
            g_clear_error(err);
            *flags |= MIO_F_CTL_SOURCECLOSE;
            ok = TRUE;
        } else {
            /* die on actual error */
            *flags |= MIO_F_CTL_ERROR;
        }
    }

    return ok;
}


int main (
    int                 argc,
    char                *argv[]) {

    GError              *err = NULL;
    ytContext_t         yx;
    MIOSource           source;
    MIOSink             sink;
    MIOAppDriver        adrv;
    uint32_t            miodflags;
    int                 rv = 0;

    /* parse options */
    ytParseOptions(&argc, &argv);

    /* check to make sure the data structures are sane */
    yfAlignmentCheck();

    /* set up logging */
    if (!logc_setup(&err)) {
        air_opterr("%s", err->message);
    }

    /* fork if necessary */
    if (!daec_setup(&err)) {
        air_opterr("%s", err->message);
    }

    /* initialize MIO option flags */
    miodflags = MIO_F_OPT_SINKLINK;

    if (!mio_config_source(&source, yaft_cliflags, &miodflags, &err)) {
        air_opterr("Cannot set up input: %s", err->message);
    }

    /* set up sink */
    if (!mio_config_sink(&source, &sink, "%s.yaf.txt", yaft_cliflags,
                         &miodflags, &err)) {
        air_opterr("Cannot set up output: %s", err->message);
    }

    /* initialize yafscii context */
    yfFlowPrepare(&(yx.flow));
    yx.fbuf = NULL;

    /* set up an app driver */
    adrv.app_open_source = ytOpenSource;
    adrv.app_open_sink = NULL;
    adrv.app_close_source = NULL;
    adrv.app_close_sink = NULL;
    adrv.app_process = ytProcess;

    /* run dispatch loop */
    if (!mio_dispatch_loop(&source, &sink, &adrv, &yx, miodflags, mio_ov_poll,
                           1, mio_ov_poll))
    {
        rv = 1;
    }

    if (yx.fbuf) {
        fBufFree(yx.fbuf);
    }

    g_message("yafscii terminating with %d flows read", yaft_flows);

    return rv;
}
